nav2_graceful_controller/src/graceful_controller.cpp

```diff
@@ -360,6 +414,48 @@ bool GracefulController::inCollision(const double & x, const double & y, const d
   return false;
 }
 
+void GracefulController::computeDistanceAlongPath(
+  const std::vector<geometry_msgs::msg::PoseStamped> & poses,
+  std::vector<double> & distances)
+{
+  distances.resize(poses.size());
+  // Do the first pose from robot
+  double d = std::hypot(poses[0].pose.position.x, poses[0].pose.position.y);
+  distances[0] = d;
+  // Compute remaining poses
+  for (size_t i = 1; i < poses.size(); ++i) {
+    d += nav2_util::geometry_utils::euclidean_distance(poses[i - 1].pose, poses[i].pose);
+    distances[i] = d;
+  }
+}
+
+std::vector<geometry_msgs::msg::PoseStamped> GracefulController::addOrientations(
+  const std::vector<geometry_msgs::msg::PoseStamped> & path)
```

Comment:
I should probably add a comment here - that we don't do anything special for reversing - because it will be handled up in L227-232